rfcs/0136-scope-expression-registration.md

```diff
@@ -0,0 +1,129 @@
+
+It is currently possible to figure some subset of this out with some background knowledge of the
+platform and docs/code reading, but is in general too difficult. In addition, there are
+workers/services that do not ever publish the scopes that they use
+for access control, making this task yet more difficult.
+
+This issue could be partially solved just by building easier ways for services/workers to publish docs
+but it will provide no guarantee that any docs published this way are valid or current. We
+should build a system that allows for these to be positively asserted.
+
+Protecting the registration system by scopes and making it the only way to do authorization allows
+a cluster admin to pick and choose which services will interact with their cluster and also allow
+them to know at any given time which services exist in their cluster.
+
+In addition, this will have the benefits of moving us to doing authorization in the auth service and
+ensure that this move is performant as this will drastically reduce the amount of bits we need to
+shovel back and forth over the wire than would be true otherwise.
+
+# Details
+
+We will define the following:
+
+* **scope expression:** A way of building complex assertions from basic scopes. This is currently defined
+  in taskcluster-lib-scopes. It will need to be standardized more generally for this to be an external api.
+* **scope expression template:** A method of building a scope expression given certain parameters. Until
+  the parameters are substituted, this is not a valid scope expression that we can use for authorization.
+* **term:** One of the parameters of a scope expression template. These are called terms when they have
+  additional metadata provided to allow for useful documentation and inspection.
+* **action:** Some functionality that is protected by scopes and requires authorization to perform. In a
+  service, these map exactly to the list of endpoints. In a worker, there is only one action and it is
+  executing a task.
+  Note: We may want to find a new name for this since we already have action tasks.
+* **principal:** An entity that attaches meaning to scopes. This can be a service, worker, or other entity
+  that might want to authorize an action or assign scopes. (i.e. taskcluster-github
+  is a principal by both authorizing some actions based on scopes that begin with "github" and also giving
+  "repo" scopes to tasks.
+* **scope namespace:** A prefix of scope strings that is owned by a principal
+
+This scope expression template registration will be implemented roughly as follows (skip to end for example):
+
+1. We will add a new field to our service api definitions at all of the cluster, service, and endpoint levels
+   that allows us to define terms at their most appropriate level. No overriding will be permitted and each
+   term must be defined with both human language documentation and a regex defining valid values for parameters
+   submitted for these terms.
```

Comment:
How will we guarantee that the human readable language matches the regex?  Defining security things twice, especially in different forms feels less than ideal.  What if we changed the format to make it easy to generate a human readable string based on the non-regex declaration?

Reply:
I think the regex and the description give different information: acceptable values, vs. "meaning".  

For example, the term "hookId" might be defined as "A user-supplied identifier for a single hook, unique within the scope of an associated `hookGroupId`".  The regex would be  `/^([a-zA-Z0-9-_]*)$/`.

Reply:
> I think the regex and the description give different information: acceptable values, vs. "meaning".

Yes, I think it's important to document the intention behind the regexp because that will usually make it easier to do a proper review and validation.